File: ZigZag_conversion/c++/ZigZag_conversion.cpp
//
//  ZigZag_conversion.cpp
//  
//
//  Created by Wen-Ting Wang on 2018/11/23.
//

#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
    string convert(string s, int numRows) {
        if(s.size() <= 1 || numRows == 1) return s;
        string ans;
        
        int step = 2*(numRows-1);
        
        for(int i = 0; i < numRows; ++i){
            for(int j = i; j < s.size(); j += step){
                ans += s[j];
                int temp = j + step - 2*i;
                if(i != 0 && i != numRows - 1 && temp < s.size())
                    ans += s[temp];
                
            }
        }
        return ans;
    }
};


int main(){
    string s = "PAYPALISHIRING";
    int numRows = 3;
    
    Solution ans;
    
    cout << ans.convert(s, numRows) << endl;
    
    return 0;
}

File: ZigZag_conversion/python/ZigZag_conversion.py
class Solution:
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """

        if(len(s) < 2 or numRows == 1):
            return s

        ans = ""
        step = 2 * (numRows - 1)

        for i in range(numRows):
            j = i
            while(j < len(s)):
                ans += s[j]

                temp = j + step - 2*i
                if (i not in [0, numRows-1]) and (temp < len(s)):
                      ans += s[temp]
                j += step

        return ans


if __name__ == "__main__":
  s = "PAYPALISHIRING"
  numRows = 3

  ans = Solution()

  print(ans.convert(s, numRows))
